x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/ProgressListenableActionFutureTests.java

```diff
@@ -0,0 +1,172 @@
+                        future.onFailure(
+                            new Exception("Failed to update progress [" + t.getClass().getName() + ':' + t.getMessage() + "]")
+                        );
+                    }
+                    throw new AssertionError(t);
+                }
+            });
+        }
+
+        for (Thread thread : threads) {
+            thread.start();
+        }
+
+        final Thread progressUpdaterThread = new Thread(() -> {
+            try {
+                startLatch.await();
+                long progress = future.start;
+                while (progress < future.end) {
+                    progress += randomLongBetween(1L, Math.max(1L, future.end - progress));
+                    future.onProgress(progress);
+                }
+                future.onResponse(future.end);
+            } catch (Throwable t) {
+                logger.error("Failed to update progress", t);
+                if (future.isDone() == false) {
+                    future.onFailure(new Exception("Failed to update progress [" + t.getClass().getName() + ':' + t.getMessage() + "]"));
+                }
+                throw new AssertionError(t);
+            }
+        });
+        progressUpdaterThread.start();
+
+        startLatch.countDown();
+
+        for (Thread thread : threads) {
+            thread.join();
+        }
+        progressUpdaterThread.join();
+        assertTrue(future.isDone());
+    }
+
+    public void testListenerCalledImmediatelyAfterResponse() throws Exception {
+        final ProgressListenableActionFuture future = randomFuture();
+        future.onResponse(randomLongBetween(future.start, future.end));
+        assertTrue(future.isDone());
+
+        final SetOnce<Long> listenerResponse = new SetOnce<>();
+        final SetOnce<Exception> listenerFailure = new SetOnce<>();
+
+        future.addListener(ActionListener.wrap(listenerResponse::set, listenerFailure::set), randomLongBetween(future.start, future.end));
+
+        assertThat(listenerResponse.get(), equalTo(future.get()));
+        assertThat(listenerFailure.get(), nullValue());
+    }
+
+    public void testListenerCalledImmediatelyAfterFailure() {
+        final ProgressListenableActionFuture future = randomFuture();
+
+        final Exception failure = new ElasticsearchException("simulated");
+        future.onFailure(failure);
+        assertTrue(future.isDone());
+
+        final SetOnce<Exception> listenerFailure = new SetOnce<>();
+        final SetOnce<Long> listenerResponse = new SetOnce<>();
+
+        future.addListener(ActionListener.wrap(listenerResponse::set, listenerFailure::set), randomLongBetween(future.start, future.end));
+
+        assertThat(listenerFailure.get(), sameInstance(failure));
+        assertThat(listenerResponse.get(), nullValue());
+    }
+
+    private static ProgressListenableActionFuture randomFuture() {
+        final long delta = randomLongBetween(1L, ByteSizeUnit.TB.toBytes(randomIntBetween(1, 10)));
```

Comment:
Don't think we need a random limit on the randomness here; also 1TB should be enough for anyone :)

```suggestion
        final long delta = randomLongBetween(1L, ByteSizeUnit.TB.toBytes(1));
```

Reply:
Agreed :)